src/transformStateWithClones.js

```diff
@@ -7,7 +7,49 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateClone = { ...state };
+  const clones = [];
+
+  for (const { type, extraData, keysToRemove } of actions) {
+    switch (type) {
+      case 'addProperties':
+        addProperties(stateClone, extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(stateClone, keysToRemove);
+        break;
+
+      case 'clear':
+        clear(stateClone);
+        break;
+
+      default:
+        return;
+    }
+
+    clones.push({ ...stateClone });
+  }
+
+  return clones;
+}
+
+function addProperties(obj, properties) {
+  for (const property in properties) {
+    obj[property] = properties[property];
+  }
+}
+
+function removeProperties(obj, keys) {
+  for (const key of keys) {
+    delete obj[key];
+  }
+}
+
+function clear(obj) {
+  for (const key in obj) {
+    delete obj[key];
+  }
 }
```

Comment:
If you use removeProperties to clear an object, you don't need this function. It duplicates the action of the previous one✨